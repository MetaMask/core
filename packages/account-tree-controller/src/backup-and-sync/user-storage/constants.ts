export const USER_STORAGE_FEATURE_PREFIX = 'multichain_accounts';

export const USER_STORAGE_WALLETS_FEATURE_KEY = `${USER_STORAGE_FEATURE_PREFIX}_wallets`;
export const USER_STORAGE_WALLETS_FEATURE_ENTRY_KEY = 'wallet';

export const USER_STORAGE_GROUPS_FEATURE_KEY = `${USER_STORAGE_FEATURE_PREFIX}_groups`;
