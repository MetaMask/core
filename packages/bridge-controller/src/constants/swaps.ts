export const SWAPS_API_V2_BASE_URL = 'https://swap.api.cx.metamask.io';
