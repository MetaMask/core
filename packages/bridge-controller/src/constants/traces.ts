export enum TraceName {
  BridgeQuotesFetched = 'Bridge Quotes Fetched',
  SwapQuotesFetched = 'Swap Quotes Fetched',
}
