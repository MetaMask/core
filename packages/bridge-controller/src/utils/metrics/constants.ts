export const UNIFIED_SWAP_BRIDGE_EVENT_CATEGORY = 'Unified SwapBridge';

/**
 * These event names map to events defined in the segment-schema: https://github.com/Consensys/segment-schema/tree/main/libraries/events/metamask-cross-chain-swaps
 */
export enum UnifiedSwapBridgeEventName {
  ButtonClicked = `${UNIFIED_SWAP_BRIDGE_EVENT_CATEGORY} Button Clicked`,
  PageViewed = `${UNIFIED_SWAP_BRIDGE_EVENT_CATEGORY} Page Viewed`,
  InputChanged = `${UNIFIED_SWAP_BRIDGE_EVENT_CATEGORY} Input Changed`,
  InputSourceDestinationFlipped = `${UNIFIED_SWAP_BRIDGE_EVENT_CATEGORY} Source Destination Flipped`,
  QuotesRequested = `${UNIFIED_SWAP_BRIDGE_EVENT_CATEGORY} Quotes Requested`,
  QuotesReceived = `${UNIFIED_SWAP_BRIDGE_EVENT_CATEGORY} Quotes Received`,
  QuoteError = `${UNIFIED_SWAP_BRIDGE_EVENT_CATEGORY} Quote Error`,
  SnapConfirmationViewed = `${UNIFIED_SWAP_BRIDGE_EVENT_CATEGORY} Snap Confirmation Page Viewed`,
  Submitted = `${UNIFIED_SWAP_BRIDGE_EVENT_CATEGORY} Submitted`,
  Completed = `${UNIFIED_SWAP_BRIDGE_EVENT_CATEGORY} Completed`,
  Failed = `${UNIFIED_SWAP_BRIDGE_EVENT_CATEGORY} Failed`,
  AllQuotesOpened = `${UNIFIED_SWAP_BRIDGE_EVENT_CATEGORY} All Quotes Opened`,
  AllQuotesSorted = `${UNIFIED_SWAP_BRIDGE_EVENT_CATEGORY} All Quotes Sorted`,
  QuoteSelected = `${UNIFIED_SWAP_BRIDGE_EVENT_CATEGORY} Quote Selected`,
}

/**
 * @deprecated remove this event property
 */
export enum MetaMetricsSwapsEventSource {
  MainView = 'Main View',
  TokenView = 'Token View',
}

export enum MetricsActionType {
  CROSSCHAIN_V1 = 'crosschain-v1',
  SWAPBRIDGE_V1 = 'swapbridge-v1',
}

export enum MetricsSwapType {
  SINGLE = 'single chain',
  CROSSCHAIN = 'crosschain',
}
