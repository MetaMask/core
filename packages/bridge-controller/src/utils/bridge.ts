import { AddressZero } from '@ethersproject/constants';
import { Contract } from '@ethersproject/contracts';
import { SolScope } from '@metamask/keyring-api';
import { abiERC20 } from '@metamask/metamask-eth-abis';
import type { CaipAssetType, CaipChainId } from '@metamask/utils';
import { isCaipChainId, isStrictHexString, type Hex } from '@metamask/utils';

import {
  formatChainIdToCaip,
  formatChainIdToDec,
  formatChainIdToHex,
} from './caip-formatters';
import {
  DEFAULT_BRIDGE_CONTROLLER_STATE,
  ETH_USDT_ADDRESS,
  METABRIDGE_ETHEREUM_ADDRESS,
} from '../constants/bridge';
import { CHAIN_IDS } from '../constants/chains';
import {
  SWAPS_CHAINID_DEFAULT_TOKEN_MAP,
  SYMBOL_TO_SLIP44_MAP,
  type SupportedSwapsNativeCurrencySymbols,
} from '../constants/tokens';
import type { BridgeAsset, BridgeControllerState } from '../types';
import { ChainId } from '../types';

export const getDefaultBridgeControllerState = (): BridgeControllerState => {
  return DEFAULT_BRIDGE_CONTROLLER_STATE;
};

/**
 * Returns the native assetType for a given chainId and native currency symbol
 * Note that the return value is used as the assetId although it is a CaipAssetType
 *
 * @param chainId - The chainId to get the native assetType for
 * @param nativeCurrencySymbol - The native currency symbol for the given chainId
 * @returns The native assetType for the given chainId
 */
const getNativeAssetCaipAssetType = (
  chainId: CaipChainId,
  nativeCurrencySymbol: SupportedSwapsNativeCurrencySymbols,
): CaipAssetType => {
  return `${formatChainIdToCaip(chainId)}/${SYMBOL_TO_SLIP44_MAP[nativeCurrencySymbol]}`;
};

/**
 * Returns the native swaps or bridge asset for a given chainId
 *
 * @param chainId - The chainId to get the default token for
 * @returns The native asset for the given chainId
 * @throws If no native asset is defined for the given chainId
 */
export const getNativeAssetForChainId = (
  chainId: string | number | Hex | CaipChainId,
): BridgeAsset => {
  const chainIdInCaip = formatChainIdToCaip(chainId);
  const nativeToken =
    SWAPS_CHAINID_DEFAULT_TOKEN_MAP[
      formatChainIdToCaip(
        chainId,
      ) as keyof typeof SWAPS_CHAINID_DEFAULT_TOKEN_MAP
    ] ??
    SWAPS_CHAINID_DEFAULT_TOKEN_MAP[
      formatChainIdToHex(
        chainId,
      ) as keyof typeof SWAPS_CHAINID_DEFAULT_TOKEN_MAP
    ];

  if (!nativeToken) {
    throw new Error(
      `No XChain Swaps native asset found for chainId: ${chainId}`,
    );
  }

  return {
    ...nativeToken,
    chainId: formatChainIdToDec(chainId),
    assetId: getNativeAssetCaipAssetType(chainIdInCaip, nativeToken.symbol),
  };
};

/**
 * A function to return the txParam data for setting allowance to 0 for USDT on Ethereum
 *
 * @returns The txParam data that will reset allowance to 0, combine it with the approval tx params received from Bridge API
 */
export const getEthUsdtResetData = () => {
  const UsdtContractInterface = new Contract(ETH_USDT_ADDRESS, abiERC20)
    .interface;
  const data = UsdtContractInterface.encodeFunctionData('approve', [
    METABRIDGE_ETHEREUM_ADDRESS,
    '0',
  ]);

  return data;
};

export const isEthUsdt = (chainId: Hex, address: string) =>
  chainId === CHAIN_IDS.MAINNET &&
  address.toLowerCase() === ETH_USDT_ADDRESS.toLowerCase();

export const sumHexes = (...hexStrings: string[]): Hex => {
  if (hexStrings.length === 0) {
    return '0x0';
  }

  const sum = hexStrings.reduce((acc, hex) => acc + BigInt(hex), BigInt(0));
  return `0x${sum.toString(16)}`;
};

/**
 * Checks whether the provided address is strictly equal to the address for
 * the default swaps token of the provided chain.
 *
 * @param address - The string to compare to the default token address
 * @param chainId - The hex encoded chain ID of the default swaps token to check
 * @returns Whether the address is the provided chain's default token address
 */
export const isSwapsDefaultTokenAddress = (
  address: string,
  chainId: Hex | CaipChainId,
) => {
  if (!address || !chainId) {
    return false;
  }

  return address === getNativeAssetForChainId(chainId)?.address;
};

/**
 * Checks whether the provided symbol is strictly equal to the symbol for
 * the default swaps token of the provided chain.
 *
 * @param symbol - The string to compare to the default token symbol
 * @param chainId - The hex encoded chain ID of the default swaps token to check
 * @returns Whether the symbol is the provided chain's default token symbol
 */
export const isSwapsDefaultTokenSymbol = (
  symbol: string,
  chainId: Hex | CaipChainId,
) => {
  if (!symbol || !chainId) {
    return false;
  }

  return symbol === getNativeAssetForChainId(chainId)?.symbol;
};

/**
 * Checks whether the address is a native asset in any supported xchain swaps network
 *
 * @param address - The address to check
 * @returns Whether the address is a native asset
 */
export const isNativeAddress = (address?: string | null) =>
  address === AddressZero || // bridge and swap apis set the native asset address to zero
  address === '' || // assets controllers set the native asset address to an empty string
  !address ||
  address.endsWith('11111111111111111111111111111111') || // token-api and bridge-api use this as the solana native assetId
  [getNativeAssetForChainId(ChainId.SOLANA).assetId].some(
    (assetId) => assetId.includes(address) && !isStrictHexString(address),
  ); // solana native assetId used in the extension client

/**
 * Checks whether the chainId matches Solana in CaipChainId or number format
 *
 * @param chainId - The chainId to check
 * @returns Whether the chainId is Solana
 */
export const isSolanaChainId = (
  chainId: Hex | number | CaipChainId | string,
) => {
  if (isCaipChainId(chainId)) {
    return chainId === SolScope.Mainnet.toString();
  }
  return chainId.toString() === ChainId.SOLANA.toString();
};
