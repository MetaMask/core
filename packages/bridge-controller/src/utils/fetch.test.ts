import { AddressZero } from '@ethersproject/constants';
import type { CaipAssetType } from '@metamask/utils';

import {
  fetchBridgeFeatureFlags,
  fetchBridgeQuotes,
  fetchBridgeTokens,
  fetchAssetPrices,
} from './fetch';
import mockBridgeQuotesErc20Erc20 from '../../tests/mock-quotes-erc20-erc20.json';
import mockBridgeQuotesNativeErc20 from '../../tests/mock-quotes-native-erc20.json';
import { BridgeClientId, BRIDGE_PROD_API_BASE_URL } from '../constants/bridge';

const mockFetchFn = jest.fn();

describe('fetch', () => {
  describe('fetchBridgeFeatureFlags', () => {
    it('should fetch bridge feature flags successfully', async () => {
      const commonResponse = {
        refreshRate: 3,
        maxRefreshCount: 1,
        support: true,
        chains: {
          '1': {
            isActiveSrc: true,
            isActiveDest: true,
          },
          '10': {
            isActiveSrc: true,
            isActiveDest: false,
          },
          '59144': {
            isActiveSrc: true,
            isActiveDest: true,
          },
          '120': {
            isActiveSrc: true,
            isActiveDest: false,
          },
          '137': {
            isActiveSrc: false,
            isActiveDest: true,
          },
          '11111': {
            isActiveSrc: false,
            isActiveDest: true,
          },
          '1151111081099710': {
            isActiveSrc: true,
            isActiveDest: true,
          },
        },
      };
      const mockResponse = {
        'extension-config': commonResponse,
        'mobile-config': commonResponse,
      };

      mockFetchFn.mockResolvedValue(mockResponse);

      const result = await fetchBridgeFeatureFlags(
        BridgeClientId.EXTENSION,
        mockFetchFn,
        BRIDGE_PROD_API_BASE_URL,
      );

      expect(mockFetchFn).toHaveBeenCalledWith(
        'https://bridge.api.cx.metamask.io/getAllFeatureFlags',
        {
          headers: { 'X-Client-Id': 'extension' },
          cacheOptions: {
            cacheRefreshTime: 600000,
          },
          functionName: 'fetchBridgeFeatureFlags',
        },
      );

      const commonExpected = {
        maxRefreshCount: 1,
        refreshRate: 3,
        support: true,
        chains: {
          'eip155:1': {
            isActiveDest: true,
            isActiveSrc: true,
          },
          'eip155:10': {
            isActiveDest: false,
            isActiveSrc: true,
          },
          'eip155:11111': {
            isActiveDest: true,
            isActiveSrc: false,
          },
          'eip155:120': {
            isActiveDest: false,
            isActiveSrc: true,
          },
          'eip155:137': {
            isActiveDest: true,
            isActiveSrc: false,
          },
          'eip155:59144': {
            isActiveDest: true,
            isActiveSrc: true,
          },
          'solana:5eykt4UsFv8P8NJdTREpY1vzqKqZKvdp': {
            isActiveDest: true,
            isActiveSrc: true,
          },
        },
      };

      expect(result).toStrictEqual({
        extensionConfig: commonExpected,
        mobileConfig: commonExpected,
      });
    });

    it('should use fallback bridge feature flags if response is unexpected', async () => {
      const commonResponse = {
        refreshRate: 3,
        maxRefreshCount: 1,
        support: 25,
        chains: {
          a: {
            isActiveSrc: 1,
            isActiveDest: 'test',
          },
          '2': {
            isActiveSrc: 'test',
            isActiveDest: 2,
          },
        },
      };
      const mockResponse = {
        'extension-config': commonResponse,
        'mobile-config': commonResponse,
      };

      mockFetchFn.mockResolvedValue(mockResponse);

      const result = await fetchBridgeFeatureFlags(
        BridgeClientId.EXTENSION,
        mockFetchFn,
        BRIDGE_PROD_API_BASE_URL,
      );

      expect(mockFetchFn).toHaveBeenCalledWith(
        'https://bridge.api.cx.metamask.io/getAllFeatureFlags',
        {
          cacheOptions: {
            cacheRefreshTime: 600000,
          },
          functionName: 'fetchBridgeFeatureFlags',
          headers: { 'X-Client-Id': 'extension' },
        },
      );

      const commonExpected = {
        maxRefreshCount: 5,
        refreshRate: 30000,
        support: false,
        chains: {},
      };
      expect(result).toStrictEqual({
        extensionConfig: commonExpected,
        mobileConfig: commonExpected,
      });
    });

    it('should handle fetch error', async () => {
      const mockError = new Error('Failed to fetch');

      mockFetchFn.mockRejectedValue(mockError);

      await expect(
        fetchBridgeFeatureFlags(
          BridgeClientId.EXTENSION,
          mockFetchFn,
          BRIDGE_PROD_API_BASE_URL,
        ),
      ).rejects.toThrow(mockError);
    });
  });

  describe('fetchBridgeTokens', () => {
    it('should fetch bridge tokens successfully', async () => {
      const mockResponse = [
        {
          address: '0x0000000000000000000000000000000000000000',
          assetId: 'eip155:10/slip44:614',
          symbol: 'ETH',
          decimals: 18,
          name: 'Ether',
          coingeckoId: 'ethereum',
          aggregators: [],
          iconUrl:
            'https://static.cx.metamask.io/api/v2/tokenIcons/assets/eip155/10/native/614.png',
          metadata: {
            honeypotStatus: {},
            isContractVerified: false,
            erc20Permit: false,
            description: {},
            createdAt: '2023-10-31T22:16:37.494Z',
          },
          chainId: 10,
        },
        {
          address: '0x1f9840a85d5af5bf1d1762f925bdaddc4201f984',
          assetId: 'eip155:10/erc20:0x1f9840a85d5af5bf1d1762f925bdaddc4201f984',
          symbol: 'ABC',
          name: 'ABC',
          decimals: 16,
          chainId: 10,
        },
        {
          address: '0x1f9840a85d5af5bf1d1762f925bdaddc4201f985',
          assetId: 'eip155:10/erc20:0x1f9840a85d5af5bf1d1762f925bdaddc4201f985',
          decimals: 16,
          chainId: 10,
        },
        {
          address: '0x1f9840a85d5af5bf1d1762f925bdaddc4201f986',
          assetId: 'eip155:10/erc20:0x1f9840a85d5af5bf1d1762f925bdaddc4201f986',
          decimals: 16,
          symbol: 'DEF',
          name: 'DEF',
          aggregators: ['lifi'],
          chainId: 10,
        },
        {
          address: '0x1f9840a85d5af5bf1d1762f925bdaddc4201f987',
          assetId: 'eip155:10/erc20:0x1f9840a85d5af5bf1d1762f925bdaddc4201f987',
          symbol: 'DEF',
          chainId: 10,
        },
        {
          address: '0x124',
          assetId: 'eip155:10/erc20:0x0b2c639c533813f4aa9d7837caf62653d097ff85',
          symbol: 'JKL',
          decimals: 16,
          chainId: 10,
        },
      ];

      mockFetchFn.mockResolvedValue(mockResponse);

      const result = await fetchBridgeTokens(
        '0xa',
        BridgeClientId.EXTENSION,
        mockFetchFn,
        BRIDGE_PROD_API_BASE_URL,
      );

      expect(mockFetchFn).toHaveBeenCalledWith(
        'https://bridge.api.cx.metamask.io/getTokens?chainId=10',
        {
          cacheOptions: {
            cacheRefreshTime: 600000,
          },
          functionName: 'fetchBridgeTokens',
          headers: { 'X-Client-Id': 'extension' },
        },
      );

      expect(result).toStrictEqual({
        '0x0000000000000000000000000000000000000000': {
          address: '0x0000000000000000000000000000000000000000',
          aggregators: [],
          assetId: 'eip155:10/slip44:614',
          chainId: 10,
          coingeckoId: 'ethereum',
          decimals: 18,
          iconUrl:
            'https://static.cx.metamask.io/api/v2/tokenIcons/assets/eip155/10/native/614.png',
          metadata: {
            createdAt: '2023-10-31T22:16:37.494Z',
            description: {},
            erc20Permit: false,
            honeypotStatus: {},
            isContractVerified: false,
          },
          name: 'Ether',
          symbol: 'ETH',
        },
        '0x1f9840a85d5af5bf1d1762f925bdaddc4201f986': {
          address: '0x1f9840a85d5af5bf1d1762f925bdaddc4201f986',
          assetId: 'eip155:10/erc20:0x1f9840a85d5af5bf1d1762f925bdaddc4201f986',
          chainId: 10,
          decimals: 16,
          name: 'DEF',
          symbol: 'DEF',
          aggregators: ['lifi'],
        },
        '0x1f9840a85d5af5bf1d1762f925bdaddc4201f984': {
          address: '0x1f9840a85d5af5bf1d1762f925bdaddc4201f984',
          assetId: 'eip155:10/erc20:0x1f9840a85d5af5bf1d1762f925bdaddc4201f984',
          chainId: 10,
          decimals: 16,
          name: 'ABC',
          symbol: 'ABC',
        },
      });
    });

    it('should handle fetch error', async () => {
      const mockError = new Error('Failed to fetch');

      mockFetchFn.mockRejectedValue(mockError);

      await expect(
        fetchBridgeTokens(
          '0xa',
          BridgeClientId.EXTENSION,
          mockFetchFn,
          BRIDGE_PROD_API_BASE_URL,
        ),
      ).rejects.toThrow(mockError);
    });
  });

  describe('fetchBridgeQuotes', () => {
    it('should fetch bridge quotes successfully, no approvals', async () => {
      mockFetchFn.mockResolvedValue(mockBridgeQuotesNativeErc20);
      const { signal } = new AbortController();

      const result = await fetchBridgeQuotes(
        {
          walletAddress: '0x388c818ca8b9251b393131c08a736a67ccb19297',
          srcChainId: 1,
          destChainId: 10,
          srcTokenAddress: AddressZero,
          destTokenAddress: AddressZero,
          srcTokenAmount: '20000',
          slippage: 0.5,
        },
        signal,
        BridgeClientId.EXTENSION,
        mockFetchFn,
        BRIDGE_PROD_API_BASE_URL,
      );

      expect(mockFetchFn).toHaveBeenCalledWith(
        'https://bridge.api.cx.metamask.io/getQuote?walletAddress=0x388C818CA8B9251b393131C08a736A67ccB19297&destWalletAddress=0x388C818CA8B9251b393131C08a736A67ccB19297&srcChainId=1&destChainId=10&srcTokenAddress=0x0000000000000000000000000000000000000000&destTokenAddress=0x0000000000000000000000000000000000000000&srcTokenAmount=20000&insufficientBal=false&resetApproval=false&slippage=0.5',
        {
          cacheOptions: {
            cacheRefreshTime: 0,
          },
          functionName: 'fetchBridgeQuotes',
          headers: { 'X-Client-Id': 'extension' },
          signal,
        },
      );

      expect(result).toStrictEqual(mockBridgeQuotesNativeErc20);
    });

    it('should fetch bridge quotes successfully, with approvals', async () => {
      mockFetchFn.mockResolvedValue([
        ...mockBridgeQuotesErc20Erc20,
        { ...mockBridgeQuotesErc20Erc20[0], approval: null },
        { ...mockBridgeQuotesErc20Erc20[0], trade: null },
      ]);
      const { signal } = new AbortController();

      const result = await fetchBridgeQuotes(
        {
          walletAddress: '0x388c818ca8b9251b393131c08a736a67ccb19297',
          srcChainId: 1,
          destChainId: 10,
          srcTokenAddress: AddressZero,
          destTokenAddress: AddressZero,
          srcTokenAmount: '20000',
          slippage: 0.5,
        },
        signal,
        BridgeClientId.EXTENSION,
        mockFetchFn,
        BRIDGE_PROD_API_BASE_URL,
      );

      expect(mockFetchFn).toHaveBeenCalledWith(
        'https://bridge.api.cx.metamask.io/getQuote?walletAddress=0x388C818CA8B9251b393131C08a736A67ccB19297&destWalletAddress=0x388C818CA8B9251b393131C08a736A67ccB19297&srcChainId=1&destChainId=10&srcTokenAddress=0x0000000000000000000000000000000000000000&destTokenAddress=0x0000000000000000000000000000000000000000&srcTokenAmount=20000&insufficientBal=false&resetApproval=false&slippage=0.5',
        {
          cacheOptions: {
            cacheRefreshTime: 0,
          },
          functionName: 'fetchBridgeQuotes',
          headers: { 'X-Client-Id': 'extension' },
          signal,
        },
      );

      expect(result).toStrictEqual(mockBridgeQuotesErc20Erc20);
    });

    it('should filter out malformed bridge quotes', async () => {
      mockFetchFn.mockResolvedValue([
        ...mockBridgeQuotesErc20Erc20,
        ...mockBridgeQuotesErc20Erc20.map(
          ({ quote, ...restOfQuote }) => restOfQuote,
        ),
        {
          ...mockBridgeQuotesErc20Erc20[0],
          quote: {
            srcAsset: {
              ...mockBridgeQuotesErc20Erc20[0].quote.srcAsset,
              decimals: undefined,
            },
          },
        },
        {
          ...mockBridgeQuotesErc20Erc20[1],
          quote: {
            srcAsset: {
              ...mockBridgeQuotesErc20Erc20[1].quote.destAsset,
              address: undefined,
            },
          },
        },
      ]);
      const { signal } = new AbortController();

      const result = await fetchBridgeQuotes(
        {
          walletAddress: '0x388c818ca8b9251b393131c08a736a67ccb19297',
          srcChainId: 1,
          destChainId: 10,
          srcTokenAddress: AddressZero,
          destTokenAddress: AddressZero,
          srcTokenAmount: '20000',
          slippage: 0.5,
        },
        signal,
        BridgeClientId.EXTENSION,
        mockFetchFn,
        BRIDGE_PROD_API_BASE_URL,
      );

      expect(mockFetchFn).toHaveBeenCalledWith(
        'https://bridge.api.cx.metamask.io/getQuote?walletAddress=0x388C818CA8B9251b393131C08a736A67ccB19297&destWalletAddress=0x388C818CA8B9251b393131C08a736A67ccB19297&srcChainId=1&destChainId=10&srcTokenAddress=0x0000000000000000000000000000000000000000&destTokenAddress=0x0000000000000000000000000000000000000000&srcTokenAmount=20000&insufficientBal=false&resetApproval=false&slippage=0.5',
        {
          cacheOptions: {
            cacheRefreshTime: 0,
          },
          functionName: 'fetchBridgeQuotes',
          headers: { 'X-Client-Id': 'extension' },
          signal,
        },
      );

      expect(result).toStrictEqual(mockBridgeQuotesErc20Erc20);
    });
  });

  describe('fetchAssetPrices', () => {
    beforeEach(() => {
      jest.clearAllMocks();
    });

    it('should fetch and combine prices for multiple currencies successfully', async () => {
      mockFetchFn
        .mockResolvedValueOnce({
          'eip155:1/erc20:0x123': { USD: '1.5' },
          'eip155:1/erc20:0x456': { USD: '2.5' },
        })
        .mockResolvedValueOnce({
          'eip155:1/erc20:0x123': { JPY: '1.3' },
          'eip155:1/erc20:0x456': null,
        })
        .mockResolvedValueOnce({
          'eip155:1/erc20:0x123': { EUR: '1.3' },
          'eip155:1/erc20:0x456': { EUR: '2.2' },
        });

      const request = {
        currencies: new Set(['USD', 'JPY', 'EUR']),
        baseUrl: 'https://api.example.com',
        fetchFn: mockFetchFn,
        clientId: 'test',
        assetIds: new Set([
          'eip155:1/erc20:0x123',
          'eip155:1/erc20:0x456',
        ]) as Set<CaipAssetType>,
      };

      const result = await fetchAssetPrices(request);

      expect(result).toStrictEqual({
        'eip155:1/erc20:0x123': {
          USD: '1.5',
          JPY: '1.3',
          EUR: '1.3',
        },
        'eip155:1/erc20:0x456': {
          USD: '2.5',
          EUR: '2.2',
        },
      });

      expect(mockFetchFn).toHaveBeenCalledTimes(3);
      expect(mockFetchFn).toHaveBeenCalledWith(
        'https://price.api.cx.metamask.io/v3/spot-prices?assetIds=eip155%3A1%2Ferc20%3A0x123%2Ceip155%3A1%2Ferc20%3A0x456&vsCurrency=USD',
        {
          headers: { 'X-Client-Id': 'test' },
          cacheOptions: { cacheRefreshTime: 30000 },
          functionName: 'fetchAssetExchangeRates',
        },
      );
      expect(mockFetchFn).toHaveBeenCalledWith(
        'https://price.api.cx.metamask.io/v3/spot-prices?assetIds=eip155%3A1%2Ferc20%3A0x123%2Ceip155%3A1%2Ferc20%3A0x456&vsCurrency=EUR',
        {
          headers: { 'X-Client-Id': 'test' },
          cacheOptions: { cacheRefreshTime: 30000 },
          functionName: 'fetchAssetExchangeRates',
        },
      );
    });

    it('should handle empty currencies set', async () => {
      const request = {
        currencies: new Set<string>(),
        baseUrl: 'https://api.example.com',
        fetchFn: mockFetchFn,
        clientId: 'test',
        assetIds: new Set([
          'eip155:1/erc20:0x123',
          'eip155:1/erc20:0x456',
        ]) as Set<CaipAssetType>,
      };

      const result = await fetchAssetPrices(request);

      expect(result).toStrictEqual({});
      expect(mockFetchFn).not.toHaveBeenCalled();
    });

    it('should handle failed requests for some currencies', async () => {
      mockFetchFn
        .mockResolvedValueOnce({
          'eip155:1/erc20:0x123': { USD: '1.5' },
        })
        .mockRejectedValueOnce(new Error('Failed to fetch EUR prices'));

      const request = {
        currencies: new Set(['USD', 'EUR']),
        baseUrl: 'https://api.example.com',
        fetchFn: mockFetchFn,
        clientId: 'test',
        assetIds: new Set([
          'eip155:1/erc20:0x123',
          'eip155:1/erc20:0x456',
        ]) as Set<CaipAssetType>,
      };

      const result = await fetchAssetPrices(request);

      expect(result).toStrictEqual({
        'eip155:1/erc20:0x123': {
          USD: '1.5',
        },
      });

      expect(mockFetchFn).toHaveBeenCalledTimes(2);
    });

    it('should handle all failed requests', async () => {
      mockFetchFn.mockRejectedValue(new Error('Failed to fetch prices'));

      const request = {
        currencies: new Set(['USD', 'EUR']),
        baseUrl: 'https://api.example.com',
        fetchFn: mockFetchFn,
        clientId: 'test',
        assetIds: new Set([
          'eip155:1/erc20:0x123',
          'eip155:1/erc20:0x456',
        ]) as Set<CaipAssetType>,
      };

      const result = await fetchAssetPrices(request);

      expect(result).toStrictEqual({});
      expect(mockFetchFn).toHaveBeenCalledTimes(2);
    });

    it('should merge prices for same asset from different currencies', async () => {
      mockFetchFn
        .mockResolvedValueOnce({
          'eip155:1/erc20:0x123': { USD: '1.5' },
          'eip155:1/erc20:0x456': null,
        })
        .mockResolvedValueOnce({
          'eip155:1/erc20:0x123': { GBP: '1.2' },
          'eip155:1/erc20:0x456': null,
        })
        .mockResolvedValueOnce({
          'eip155:1/erc20:0x123': { JPY: '165' },
          'eip155:1/erc20:0x456': null,
        })
        .mockResolvedValueOnce({
          'eip155:1/erc20:0x123': { EUR: '1.3' },
          'eip155:1/erc20:0x456': null,
        });

      const request = {
        currencies: new Set(['USD', 'GBP', 'JPY', 'EUR']),
        baseUrl: 'https://api.example.com',
        fetchFn: mockFetchFn,
        clientId: 'test',
        assetIds: new Set([
          'eip155:1/erc20:0x123',
          'eip155:1/erc20:0x456',
        ]) as Set<CaipAssetType>,
      };

      const result = await fetchAssetPrices(request);

      expect(result).toStrictEqual({
        'eip155:1/erc20:0x123': {
          USD: '1.5',
          GBP: '1.2',
          EUR: '1.3',
          JPY: '165',
        },
      });
    });

    it('should handle mixed successful and empty responses', async () => {
      mockFetchFn
        .mockResolvedValueOnce({
          'eip155:1/erc20:0x123': { USD: '1.5' },
        })
        .mockResolvedValueOnce({});

      const request = {
        currencies: new Set(['USD', 'EUR']),
        baseUrl: 'https://api.example.com',
        fetchFn: mockFetchFn,
        clientId: 'test',
        assetIds: new Set([
          'eip155:1/erc20:0x123',
          'eip155:1/erc20:0x456',
        ]) as Set<CaipAssetType>,
      };

      const result = await fetchAssetPrices(request);

      expect(result).toStrictEqual({
        'eip155:1/erc20:0x123': {
          USD: '1.5',
        },
      });
    });

    it('should handle malformed API responses', async () => {
      mockFetchFn
        .mockResolvedValueOnce(null)
        .mockResolvedValueOnce(undefined)
        .mockResolvedValueOnce('invalid format');

      const request = {
        currencies: new Set(['USD', 'EUR', 'GBP']),
        baseUrl: 'https://api.example.com',
        fetchFn: mockFetchFn,
        clientId: 'test',
        assetIds: new Set([
          'eip155:1/erc20:0x123',
          'eip155:1/erc20:0x456',
        ]) as Set<CaipAssetType>,
      };

      const result = await fetchAssetPrices(request);

      expect(result).toStrictEqual({});
      expect(mockFetchFn).toHaveBeenCalledTimes(3);
    });

    it('should handle empty assetIds', async () => {
      const request = {
        currencies: new Set(['USD', 'EUR', 'GBP']),
        baseUrl: 'https://api.example.com',
        fetchFn: mockFetchFn,
        clientId: 'test',
        assetIds: new Set([]) as Set<CaipAssetType>,
      };

      const result = await fetchAssetPrices(request);

      expect(result).toStrictEqual({});
      expect(mockFetchFn).toHaveBeenCalledTimes(0);
    });

    it('should handle network errors with appropriate status codes', async () => {
      mockFetchFn
        .mockRejectedValueOnce(new Error('404 Not Found'))
        .mockRejectedValueOnce(new Error('500 Internal Server Error'))
        .mockRejectedValueOnce(new Error('Network Error'));

      const request = {
        currencies: new Set(['USD', 'EUR', 'GBP']),
        baseUrl: 'https://api.example.com',
        fetchFn: mockFetchFn,
        clientId: 'test',
        assetIds: new Set([
          'eip155:1/erc20:0x123',
          'eip155:1/erc20:0x456',
        ]) as Set<CaipAssetType>,
      };

      const result = await fetchAssetPrices(request);

      expect(result).toStrictEqual({});
      expect(mockFetchFn).toHaveBeenCalledTimes(3);
    });
  });
});
