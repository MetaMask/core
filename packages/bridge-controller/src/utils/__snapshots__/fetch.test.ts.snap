// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fetch fetchBridgeQuotes should filter out malformed bridge quotes 1`] = `
Array [
  Array [
    "Quote validation failed",
    Object {
      "socket": Set {
        "quote.requestId",
        "quote.srcChainId",
        "quote.srcAsset.decimals",
        "quote.srcTokenAmount",
        "quote.destChainId",
        "quote.destAsset",
        "quote.destTokenAmount",
        "quote.feeData",
        "quote.bridges",
        "quote.steps",
        "quote.srcAsset",
        "quote.destAsset.address",
      },
      "undefined": Set {
        "quote",
      },
    },
  ],
]
`;
